jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/AsyncJSON.java

```diff
@@ -0,0 +1,1389 @@
+    private final NumberBuilder numberBuilder = new NumberBuilder();
+    private final Utf8StringBuilder stringBuilder = new Utf8StringBuilder(32);
+    private final Factory factory;
+    private List<ByteBuffer> chunks;
+
+    public AsyncJSON(Factory factory)
+    {
+        this.factory = factory;
+    }
+
+    // Used by tests only.
+    boolean isEmpty()
+    {
+        return stack.isEmpty();
+    }
+
+    /**
+     * <p>Feeds the parser with the given bytes chunk.</p>
+     *
+     * @param bytes the bytes to parse
+     * @return whether the JSON parsing was complete
+     * @throws IllegalArgumentException if the JSON is malformed
+     */
+    public boolean parse(byte[] bytes)
+    {
+        return parse(bytes, 0, bytes.length);
+    }
+
+    /**
+     * <p>Feeds the parser with the given bytes chunk.</p>
+     *
+     * @param bytes the bytes to parse
+     * @param offset the offset to start parsing from
+     * @param length the number of bytes to parse
+     * @return whether the JSON parsing was complete
+     * @throws IllegalArgumentException if the JSON is malformed
+     */
+    public boolean parse(byte[] bytes, int offset, int length)
+    {
+        return parse(ByteBuffer.wrap(bytes, offset, length));
+    }
+
+    /**
+     * <p>Feeds the parser with the given buffer chunk.</p>
+     *
+     * @param buffer the buffer to parse
+     * @return whether the JSON parsing was complete
+     * @throws IllegalArgumentException if the JSON is malformed
+     */
+    public boolean parse(ByteBuffer buffer)
+    {
+        try
+        {
+            if (factory.isDetailedParseException())
+            {
+                if (chunks == null)
+                    chunks = new ArrayList<>();
+                ByteBuffer copy = buffer.isDirect()
+                    ? ByteBuffer.allocateDirect(buffer.remaining())
+                    : ByteBuffer.allocate(buffer.remaining());
+                copy.put(buffer).flip();
+                chunks.add(copy);
+                buffer.flip();
+            }
+
+            if (stack.isEmpty())
+                stack.push(State.COMPLETE, UNSET);
+
+            while (true)
+            {
+                Frame frame = stack.peek();
+                State state = frame.state;
+                switch (state)
+                {
+                    case COMPLETE:
+                    {
+                        if (frame.value == UNSET)
+                        {
+                            if (parseAny(buffer))
+                                break;
+                            return false;
+                        }
+                        else
+                        {
+                            while (buffer.hasRemaining())
+                            {
+                                int position = buffer.position();
+                                byte peek = buffer.get(position);
+                                if (isWhitespace(peek))
```

Comment:
If this check is outside of the JSON document, then this version of `isWhitespace` is too restrictive. right?

The `isWhitespace(byte)` method is valid only for within a JSON document.

Reply:
This is a JSON document that must be delimited by other means, e.g. HTTP framing or WebSocket message framing.
As such it is part of the JSON document until `complete()` is called.